### PR TITLE
RHCLOUD-41014: Add CalculateTuples to resource_service

### PR DESCRIPTION
**Pull request by Adam0Brien**
## Describe your changes

- Calculate tuples to create and delete based off of the version in the DB

Test results:
```sh
=== RUN   TestCalculateTuples
=== RUN   TestCalculateTuples/version_0_creates_initial_tuple
INFO msg=CalculateTuples called - version: 0, key: {localResourceID:test-resource resourceType:host reporter:{reporterType:HBI reporterInstanceId:test-instance}}
INFO msg=CREATE scenario - version 0, processing initial tuple
INFO msg=Found workspace ID for version 0 - current: 'workspace-initial'
INFO msg=Created tuple to create: host:test-resource#workspace@rbac:workspace:workspace-initial
=== RUN   TestCalculateTuples/workspace_change_creates_and_deletes_tuples
INFO msg=CalculateTuples called - version: 2, key: {localResourceID:test-resource resourceType:host reporter:{reporterType:HBI reporterInstanceId:test-instance}}
INFO msg=Found workspace IDs - current: 'workspace-new', previous: 'workspace-old'
INFO msg=Created tuple to create: host:test-resource#workspace@rbac:workspace:workspace-new
INFO msg=Created tuple to delete: host:test-resource#workspace@rbac:workspace:workspace-old
=== RUN   TestCalculateTuples/workspace_change_creates_and_deletes_tuples_version_1
INFO msg=CalculateTuples called - version: 1, key: {localResourceID:test-resource resourceType:host reporter:{reporterType:HBI reporterInstanceId:test-instance}}
INFO msg=Found workspace IDs - current: 'workspace-new', previous: 'workspace-old'
INFO msg=Created tuple to create: host:test-resource#workspace@rbac:workspace:workspace-new
INFO msg=Created tuple to delete: host:test-resource#workspace@rbac:workspace:workspace-old
=== RUN   TestCalculateTuples/same_workspace_creates_only
INFO msg=CalculateTuples called - version: 2, key: {localResourceID:test-resource resourceType:host reporter:{reporterType:HBI reporterInstanceId:test-instance}}
INFO msg=Found workspace IDs - current: 'workspace-same', previous: 'workspace-same'
INFO msg=Created tuple to create: host:test-resource#workspace@rbac:workspace:workspace-same
--- PASS: TestCalculateTuples (0.00s)
    --- PASS: TestCalculateTuples/version_0_creates_initial_tuple (0.00s)
    --- PASS: TestCalculateTuples/workspace_change_creates_and_deletes_tuples (0.00s)
    --- PASS: TestCalculateTuples/workspace_change_creates_and_deletes_tuples_version_1 (0.00s)
    --- PASS: TestCalculateTuples/same_workspace_creates_only (0.00s)
    
   ```

## Ticket reference (if applicable)
Fixes #

## Checklist

* [ ] Are the agreed upon acceptance criteria fulfilled?

* [ ] Was the 4-eye-principle applied? (async PR review, pairing, ensembling)

* [ ] Do your changes have passing automated tests and sufficient observability?

* [ ] Are the work steps you introduced repeatable by others, either through automation or documentation?
  * [ ] If automation is possible but not done due to other constraints, a ticket to the tech debt sprint is added
  * [ ] An SOP (Standard Operating Procedure) was created

* [ ] The Changes were automatically built, tested, and  - if needed, behind a feature flag - deployed to our production environment. (**Please check this when the new deployment is done and you could verify it.**)

* [ ] Are the agreed upon coding/architectural practices applied?

* [ ] Are security needs fullfilled? (e.g. no internal URL)

* [ ] Is the corresponding Ticket in the right state? (should be on "review" now, put to done when this change made it to production)

* [ ] For changes to the public API / code dependencies: Was the whole team (or a sufficient amount of ppl) able to review?

## Summary by Sourcery

Generate create and delete relation tuples based on workspace_id version history by extending the repository API and adding a new CalculateTuples usecase with supporting model types and comprehensive tests

New Features:
- Add CalculateTuples method to resource_service for generating RBAC tuples based on workspace_id changes
- Introduce TupleEvent, RelationsTuple, and TuplesToReplicate model types for tuple processing
- Extend ResourceRepository interface with FindVersionedRepresentationsByVersion to fetch current and previous workspace representations
- Implement FindVersionedRepresentationsByVersion in both real and fake repositories

Tests:
- Add unit tests for CalculateTuples covering initial creation, workspace-change create/delete, and no-op scenarios
- Add contract and edge-case tests for FindVersionedRepresentationsByVersion including version zero, large versions, and error handling

### CODE REVIEW

**Comment by Adam0Brien** [comment]
Originally made the PR in #844, Starting on a new branch

File: internal/biz/usecase/resources/resource_service.go [hunk 1]
```diff
@@ -822,3 +835,204 @@ func computeReadAfterWrite(uc *Usecase, write_visibility v1beta2.WriteVisibility
 	}
 	return !common.IsNil(uc.ListenManager) && uc.Config.ReadAfterWriteEnabled && isSPInAllowlist(reporterPrincipal, uc.Config.ReadAfterWriteAllowlist)
 }
+
+func (uc *Usecase) CalculateTuples(tuple *kessel.Relationship, currentCommonVersion model.Version, key model.ReporterResourceKey) (*kessel.RelationTupleFilter, *kessel.RelationTupleFilter) {
```
[the remark this answers is not in the document]
Comment (snehagunta) [reply internal/biz/usecase/resources/resource_service.go:839]: Also you do not need to take an incoming tuple as input, just take the currentVersion and reporterResourceKey and lookup the current tuple and previous tuple all in this method

File: internal/biz/model/common_representation.go [hunk 2]
```diff
@@ -96,3 +96,13 @@ func DeserializeCommonRepresentation(snapshot *CommonRepresentationSnapshot) Com
 func (cr CommonRepresentation) CreateSnapshot() (CommonRepresentationSnapshot, error) {
 	return cr.Serialize(), nil
 }
+
+// GetField extracts a field from the common representation data
+func (cr CommonRepresentation) GetField(fieldName string) string {
```
Comment (snehagunta) [inline internal/biz/model/common_representation.go:101]: I would add getters only if needed, if you give access to all the fields this way, we could be breaking encapsulation. You can use this utility method if you want, but make it unexported and only expose the WorkspaceId(). the principle is to hide as much as you can and expose as little as possible.

File: internal/data/resource_repository.go [hunk 3]
```diff
@@ -241,3 +248,85 @@ func (r *resourceRepository) GetDB() *gorm.DB {
...
+	// Use provided transaction or fall back to regular DB session
+	db := tx
+	if db == nil {
+		db = r.db.Session(&gorm.Session{})
+	}
+
+	query := db.Table("reporter_resources rr").
+		Select("cr.data, cr.version").
+		Joins("JOIN common_representations cr ON rr.resource_id = cr.resource_id").
+		Where("LOWER(rr.local_resource_id) = LOWER(?)", key.LocalResourceId().Serialize()).
```
Comment (snehagunta) [inline internal/data/resource_repository.go:264]: I am considering moving these where clauses to a common helper function in the repository, so we can use them in both the lookup functions

File: internal/data/resource_repository.go [hunk 4]
```diff
@@ -241,3 +248,85 @@ func (r *resourceRepository) GetDB() *gorm.DB {
 func (r *resourceRepository) GetTransactionManager() usecase.TransactionManager {
 	return r.transactionManager
 }
+
+func (r *resourceRepository) FindCommonRepresentationsByVersion(tx *gorm.DB, key bizmodel.ReporterResourceKey, currentVersion uint) ([]CommonRepresentationsByVersion, error) {
```
[the remark this answers is not in the document]
Comment (snehagunta) [reply internal/data/resource_repository.go:252]: A test I am interested in seeing is what is the result if current version is 0. This may help us eliminate the 

if create do x
if update do y
if delete do z

switch in the resource_service function

File: internal/biz/usecase/resources/resource_service.go [hunk 5]
```diff
@@ -852,3 +863,90 @@ func computeReadAfterWrite(uc *Usecase, write_visibility v1beta2.WriteVisibility
...
+func (uc *Usecase) CalculateTuples(tupleEvent model.TupleEvent) (model.TuplesToReplicate, error) {
+	currentVersion := tupleEvent.Version().Uint()
+	key := tupleEvent.ReporterResourceKey()
+
+	uc.Log.Infof("CalculateTuples called - version: %d, key: %+v", currentVersion, key)
+
+	var currentWorkspaceID, previousWorkspaceID string
+
+	// Handle create scenario (version 0)
+	if currentVersion == 0 {
```
Comment (snehagunta) [inline internal/biz/usecase/resources/resource_service.go:876]: I am interested in seeing if there is a way to avoid this if else,

File: internal/data/resource_repository.go [hunk 6]
```diff
@@ -26,6 +27,11 @@ type FindResourceByKeysResult struct {
 	ConsistencyToken      string    `gorm:"column:consistency_token"`
 }
 
+type CommonRepresentationsByVersion struct {
```
[the remark this answers is not in the document]
Comment (snehagunta) [reply internal/data/resource_repository.go:30]: Let us call it RepresentationsByVersion for now

File: internal/data/resource_repository.go [hunk 7]
```diff
@@ -241,3 +248,34 @@ func (r *resourceRepository) GetDB() *gorm.DB {
...
+		db = r.db.Session(&gorm.Session{})
+	}
+
+	query := db.Table("reporter_resources rr").
+		Select("cr.data, cr.version").
+		Joins("JOIN common_representations cr ON rr.resource_id = cr.resource_id").
+		Where("LOWER(rr.local_resource_id) = LOWER(?)", key.LocalResourceId().Serialize()).
+		Where("LOWER(rr.resource_type) = LOWER(?)", key.ResourceType().Serialize()).
+		Where("LOWER(rr.reporter_type) = LOWER(?)", key.ReporterType().Serialize()).
+		Where("rr.tombstone = ?", false).
```
Comment (snehagunta) [inline internal/data/resource_repository.go:267]: Remove tombstone for now, we will address it when we figure out deletes

File: internal/data/resource_repository.go [hunk 8]
```diff
@@ -241,3 +248,34 @@ func (r *resourceRepository) GetDB() *gorm.DB {
 func (r *resourceRepository) GetTransactionManager() usecase.TransactionManager {
 	return r.transactionManager
 }
+
+func (r *resourceRepository) FindVersionedRepresentationsByVersion(tx *gorm.DB, key bizmodel.ReporterResourceKey, currentVersion uint) ([]RepresentationsByVersion, error) {
```
Comment (snehagunta) [inline internal/data/resource_repository.go:252]: Add a TODO here to say this is focused on getting common version for now but will need to be expanded to get reporter representation version as well later

**Review by snehagunta — APPROVED** [review]
